### DISCUSSION TITLE
tesla-http-proxy-addon + Nginx proxy manager (NGM)

### DISCUSSION THREAD

**Discussion by MicheleMercuri** [llamafilm/tesla-http-proxy-addon#59] [Q&A] [answered]
Hi @ all, I think I've tried all the solutions but I couldn't get the addon to work for using the new Tesla APIs :( 
My configuration is: 
HAOS + NGINX Proxy Manager which I already use for remote access to HA using https + duckdns. 

To configure the tesla-http-proxy-addon I tried this: 

in NGINX Proxy Manager I added a second domain (first domain is used to access to HAOS remotely by https on 443 port):
NGINX Proxy Manager configuration:

(1st panel): Details
```
Domain Names *
myteslamodel3.duckdns.org
Forward Hostname / IP*: 192.168.1.100 (this is internal ip addr of my HAOS that runs on default port 8123)
Forward Port *: 4430
Websockets Support: checked
```
-------------------------------------------
(3rd panel) SSL:
```
SSL Certificate: myslamodel3.duckdns.org
Force SSL: checked
HTTP/2 Support: checked
```

**Certificate provider: Let's Encrypt**
--------------------------------------------
(4th panel): Advanced
```
        add_header Strict-Transport-Security "max-age=31536000; includeSubDomains" always;
        set $target 192.168.1.100;
        # temporary Flask app for initial auth
        location /callback {
            proxy_pass http://$target:8099;
        }
        # static public key for Tesla
        location /.well-known/appspecific/com.tesla.3p.public-key.pem {
            root /share/tesla;
            try_files /com.tesla.3p.public-key.pem =404;
        }
        location = /favicon.ico {
            log_not_found off;
        }
        location = /robots.txt {
            log_not_found off;
        }
```
in HAOS in configuration.yaml

```
http:
  use_x_forwarded_for: true
  trusted_proxies:
    - 172.30.33.0/24
```
Nothing works ☹, # static public key for Tesla is not reachable from the outside (bad gateway response), the tesla-http-proxy-addon app stops working, this’s tesla-http-proxy-addon log:
```
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
[01:33:25] INFO: Found existing keypair
[01:33:25] INFO: Testing public key...
HTTP/2 403 
server: nginx
date: Sun, 24 Mar 2024 00:33:29 GMT
content-type: text/plain; charset=utf-8
content-length: 14
referrer-policy: no-referrer
x-content-type-options: nosniff
x-frame-options: SAMEORIGIN
strict-transport-security: max-age=31536000; includeSubDomains

[01:33:29] ERROR: Fix public key before proceeding.
s6-rc: info: service legacy-services: stopping
s6-rc: info: service legacy-services successfully stopped
s6-rc: info: service legacy-cont-init: stopping
s6-rc: info: service legacy-cont-init successfully stopped
s6-rc: info: service fix-attrs: stopping
s6-rc: info: service fix-attrs successfully stopped
s6-rc: info: service s6rc-oneshot-runner: stopping
s6-rc: info: service s6rc-oneshot-runner successfully stopped
```
**Someone have any suggestions to help me to fix it? Tnks**

**Comment by MicheleMercuri** [comment]
I moved key: com.tesla.3p.public-key.pem in www/tesla an now all seems works, I can download files from domain https.... but i can't Enroll public key in vehicle. log repost an error but isn't true because I can download  com.tesla.3p.public-key.pem from https://teslamickmmm.duckdns.org/local/tesla/com.tesla.3p.public-key.pem

```
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service webui: starting
s6-rc: info: service webui successfully started
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
[18:23:35] webui:INFO: Starting Flask server for Web UI...
[18:23:35] werkzeug:INFO: WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:8099
 * Running on http://172.30.33.7:8099
[18:23:35] werkzeug:INFO: Press CTRL+C to quit
[18:23:37] INFO: Found existing keypair
[18:23:37] INFO: Testing public key...
HTTP/2 301 
server: nginx
date: Sun, 24 Mar 2024 17:23:37 GMT
content-type: text/html
content-length: 162
location: https://teslamickmmm.duckdns.org/local/tesla/com.tesla.3p.public-key.pem

<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx</center>
</body>
</html>
[18:23:37] INFO: Running auth.py
[18:23:37] auth:INFO: Generating Partner Authentication Token
[18:23:38] auth:INFO: Registering Tesla account...
[18:23:40] auth:ERROR: Error 424: {"response":null,"error":"Public key download failed for https://teslamickmmm.duckdns.org/.well-known/appspecific/com.tesla.3p.public-key.pem, error: must return 200 response code, got 301","error_description":""}
s6-rc: info: service legacy-services: stopping
s6-rc: info: service legacy-services successfully stopped
s6-rc: info: service webui: stopping
s6-rc: info: service webui successfully stopped
s6-rc: info: service legacy-cont-init: stopping
s6-rc: info: service legacy-cont-init successfully stopped
s6-rc: info: service fix-attrs: stopping
s6-rc: info: service fix-attrs successfully stopped
s6-rc: info: service s6rc-oneshot-runner: stopping
s6-rc: info: service s6rc-oneshot-runner successfully stopped
```

**Reply by llamafilm** [reply]
The error is clearly stated in the log. Your server returns 301 redirect.
The URL you wrote (https://teslamickmmm.duckdns.org/local/tesla/com.tesla.3p.public-key.pem) returns an error for me. But even if that URL worked, that doesn’t really matter because the public key must be accessible on the URL described in the docs and in the log.

**Reply by MicheleMercuri** [reply]
it seems fixed, many thanks ;) . I moved com.tesla.3p.public-key.pem in www/tesla 
then in NGINX proxy manager -->

```
location = /.well-known/appspecific/com.tesla.3p.public-key.pem {
    proxy_pass http://192.168.0.100:8123/local/tesla/com.tesla.3p.public-key.pem;
}

```
this's log:

```
s6-rc: info: service s6rc-oneshot-runner: starting
s6-rc: info: service s6rc-oneshot-runner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service webui: starting
s6-rc: info: service webui successfully started
s6-rc: info: service legacy-services: starting
s6-rc: info: service legacy-services successfully started
[22:02:07] webui:INFO: Starting Flask server for Web UI...
[22:02:07] werkzeug:INFO: WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:8099
 * Running on http://172.30.33.7:8099
[22:02:07] werkzeug:INFO: Press CTRL+C to quit
[22:02:08] INFO: Found existing keypair
[22:02:08] INFO: Testing public key...
HTTP/2 200 
server: nginx
date: Sun, 24 Mar 2024 21:02:09 GMT
content-type: application/x-x509-ca-cert
content-length: 178
cache-control: public, max-age=2678400
referrer-policy: no-referrer
x-content-type-options: nosniff
x-frame-options: SAMEORIGIN
etag: "17t179b1d138d33-b2"
last-modified: Thu, 21 Mar 2024 22:19:19 GMT
accept-ranges: bytes

-----BEGIN PUBLIC KEY-----
MFkwEwXXXXXXXXXXXXXXXXXIKoZIzXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
9b8qRXXXXXXXXXXXXXXXXXr7v9rXO0zPA2ExJBggMwY4BgK4HzxfCJ+95UAdKA==
-----END PUBLIC KEY-----
[22:02:09] INFO: Starting Tesla HTTP Proxy

Do not listen on a network interface without adding client authentication. Unauthorized clients may
be used to create excessive traffic from your IP address to Tesla's servers, which Tesla may respond
to by rate limiting or blocking your connections.
2024-03-24T22:02:09+01:00 [debug] Creating proxy
2024-03-24T22:02:09+01:00 [info ] Listening on 0.0.0.0:443
```

**Reply by gawsneakerz** [reply]
@MicheleMercuri, you saved me tonight. Was struggling with this the whole day already. Thank you!!

**Reply by MicheleMercuri** [reply]
I've been struggling with it for many hours too :) But despite the tesla-http-proxy addon seeming to work now, I can't get the "Tesla Custom Integration" config flow. I believe it's still due to an issue with NGM; could it be the selfsigned.pem not being accessible? In the config flow, it's in /share/tesla/selfsigned.pem, but I copied it into www/tesla, in I try to change with /local/tesla/ the integration don't complete config flow correctly. I have no idea how to solve this :(

this is the log, [llamafilm](https://github.com/llamafilm) do you have any suggestion to solve my problem? 
```
Registratore: homeassistant.config_entries
Fonte: config_entries.py:528
Prima occorrenza: 15:02:35 (6 occorrenze)
Ultima registrazione: 15:27:28

Config entry 'emailemail@email.it' for tesla_custom integration could not authenticate
```

**Reply by bigdog989** [reply]
I've been trying to work this out as well, and the info in this thread has been super helpful, i was just wondering if someone could explain where the com.tesla.3p.public-key.pem and selfsigned.pem files come from and where exactly they need to go?

**Reply by bigdog989** [reply]
Think I've answered my own question.

Files originate in home assistant at "/share/tesla" and need to be moved to "/homeassistant/www/tesla" for anyone else that's playing along

**Reply by hexonMD** [reply]
Mine doesn't create that file? I'm lost...

**Comment by MicheleMercuri** [comment]
I've been at a dead end for days :( . I've correctly set-up the Tesla HTTP Proxy add-on and completed the procedure for enrolling the public key in my vehicle. 
The test of the public key endpoint is positive, but I can't finish the config flow of the "Tesla Custom Integration" that stops with this error:
`Config entry 'emailemail@email.it' for tesla_custom integration could not authenticate`
My car, still has access to the **old Tesla APIs**, could this be the reason?
My configuration is as follows:

1. HAOS
2. NGINX Proxy Manager
3. A second-level domain certified with DuckDNS that I use both for remote access to the Home Assistant server and for the Tesla HTTP Proxy add-on.
4. moved in \config\www\tesla : "com.tesla.3p.public-key.pem"
5. in NGINX Proxy Manager --> advanced --> custom configurations:
```
 location = /.well-known/appspecific/com.tesla.3p.public-key.pem {
    proxy_pass http://192.168.1.190:8123/local/tesla/com.tesla.3p.public-key.pem;
 }
```
![1](https://github.com/llamafilm/tesla-http-proxy-addon/assets/97225788/74bc95ae-d136-4c70-87b1-115d28064d12)
![2](https://github.com/llamafilm/tesla-http-proxy-addon/assets/97225788/b5484334-f096-4647-97e0-4612c47ea884)
![3](https://github.com/llamafilm/tesla-http-proxy-addon/assets/97225788/2071d4e2-2aee-4bd4-9853-f75cbd35d368)

**Reply by llamafilm** [reply]
It sounds like you have the addon setup correctly and it's working fine.  Make sure you're using the latest version of the Tesla integration (update via HACS).  Then try deleting the vehicle from the devices page and adding it back again.  If that doesn't work, please open an issue against the Tesla integration, including the debug logs.

**Reply by MicheleMercuri** [reply]
many thanks [llamafilm](https://github.com/llamafilm), I will try deleting vehicle and adding again and if not works I will open an issue in Tesla integration git. A Tesla can have access to both versions of the APIs (new and old one) at the same time?

**Reply by llamafilm** [reply]
The old Owner API still works for my 2021 model 3, for reading sensors only. Commands don’t work. I expect they will disable it completely at some like.

**Reply by MicheleMercuri** [reply]
I think it's a problem of my HA. The Custom Integration can't load the SSL certificate. Maybe it's due to the configuration: tesla-proxy+NPM. I will try to find a solution
```
Registratore: teslajsonpy.controller
Fonte: custom_components/tesla_custom/__init__.py:170
integration: Tesla Custom Integration (documentazione, problemi)
Prima occorrenza: 18:15:24 (2 occorrenze)
Ultima registrazione: 18:16:39

Unable to load custom SSL certificate from http://192.168.1.190:8123/local/tesla/selfsigned.pem
Unable to load custom SSL certificate from /local/tesla/selfsigned.pem
```

**Reply by jerry34ha** [reply, marked as answer]
First, thank you @MicheleMercuri, thanks to you I could configure the add-on with NGM.

To install the Tesla Custom integration, I:
* Checked the "use fleet API" checkbox
* Entered my Tesla account email
* copied the refresh token generated by the tesla-http-proxy add-on
* the accountID was prefilled
*  changed the location of the selfsigned.pem to /config/www/tesla/selfsigned.pem (it was pre-filled to /share/tesla)

And it seemed to work as I can see my car as a device now in HA.

**Reply by MicheleMercuri** [reply]
works for me too ;). (tesla-http-proxy-addon + NGM :) )

**Reply by stefanraduionescu** [reply]
I have a similiar situation here and can't find a solution for several days.
tesla custom integration 3.24.2
ha 2024.11.1
the car is tesla model 3 2024

Tesla-http-proxy-addon + NGM (configured working fine)
Tesla http proxy is working fine (with the 2 files moved in /config/www/tesla
Vehicle connected fine (the app procedure)

I can't make it work Tesla custom integration with fleet api
I changed the location of selfigned.pem completed in config and I still have "Failed to set up. Check the logs" 
In logs i have an error about region:

2024-11-14 17:12:31.683 DEBUG (MainThread) [custom_components.tesla_custom2] Trusting CA: {'subject': ((('countryName', 'TR'),), (('localityName', 'Gebze - Kocaeli'),), (('organizationName', 'Turkiye Bilimsel ve Teknolojik Arastirma Kurumu - TUBITAK'),), (('organizationalUnitName', 'Kamu Sertifikasyon Merkezi - Kamu SM'),), (('commonName', 'TUBITAK Kamu SM SSL Kok Sertifikasi - Surum 1'),)), 'issuer': ((('countryName', 'TR'),), (('localityName', 'Gebze - Kocaeli'),), (('organizationName', 'Turkiye Bilimsel ve Teknolojik Arastirma Kurumu - TUBITAK'),), (('organizationalUnitName', 'Kamu Sertifikasyon Merkezi - Kamu SM'),), (('commonName', 'TUBITAK Kamu SM SSL Kok Sertifikasi - Surum 1'),)), 'version': 3, 'serialNumber': '01', 'notBefore': 'Nov 25 08:25:55 2013 GMT', 'notAfter': 'Oct 25 08:25:55 2043 GMT'}
2024-11-14 17:12:31.721 DEBUG (MainThread) [teslajsonpy.controller] 552 endpoints loaded
2024-11-14 17:12:31.722 DEBUG (MainThread) [teslajsonpy.connection] Token expiration in -20042 days, 6:47:29
2024-11-14 17:12:31.722 DEBUG (MainThread) [teslajsonpy.connection] Oauth expiration detected
2024-11-14 17:12:31.722 DEBUG (MainThread) [teslajsonpy.connection] Refreshing access token with refresh_token
2024-11-14 17:12:32.364 DEBUG (MainThread) [teslajsonpy.connection] Saved new auth info {'access_token': 'eyJhbGciOiJSUzI1NiIsInR5cCI6Ikp
2024-11-14 17:12:32.364 DEBUG (MainThread) [teslajsonpy.connection] Successfully refreshed oauth
2024-11-14 17:12:32.364 DEBUG (MainThread) [teslajsonpy.connection] get: https://c03d64a7-tesla-http-proxy/api/1/products {}
2024-11-14 17:12:32.578 DEBUG (MainThread) [teslajsonpy.connection] 412: {"error":"Account 99558706-0d63-446a-96dd-34dc1d1936ea must be registered in the current region https://fleet-api.prd.na.vn.cloud.tesla.com, please see https://developer.tesla.com/docs/fleet-api/endpoints/partner-endpoints#register"}

**Reply by llamafilm** [reply]
@stefanraduionescu it sounds like you chose the wrong region when setting it up. The error message shows `na` which means North America. You need to remove the Tesla integration and set it up again with EU.

**Reply by MicheleMercuri** [reply]
@llamafilm 1 question: I think it's better to move only the file com.tesla.3p.public-key.pem to \config\www\... and not selfsigned.pem for security reasons, right? Initially, I had placed both files there, but then I decided to leave only com.tesla.3p.public-key.pem. and all works fine.

**Reply by llamafilm** [reply]
Yes of course, there is no reason to expose `selfsigned.pem` publicly.  Although I can't really think of a security risk by doing so.

**Reply by stefanraduionescu** [reply]
Thank you for answers!
@llamafilm The token I got from Tesla proxy starts with NA (Nord America) but I selected Europe, tryed several times .... I don't know why I got an North America token. Should the developer app access has to do something with this ?

I removed it and added it again several times, can't find the setting for the region ...
Should i start again with the Tesla Proxy or there is other configuration for the region in Tesla custom integration? 
@MicheleMercuri I did not understand properly when is the correct moment to move the file (or files):
1. I make sure NPM is running (configured as you mentioned) 
2. Start Tesla HTTP Proxy with "Regenerate Tesla authentication", ( - Region EU in my case), client id, client secret and the domain configured in NPM for this )
3. Open web-ui - (Login / paste the url / generate token(save) then ok button ) 
4.  -this is the moment when I move the files
5. Came back to web-ui (Enroll public key in the car - green was success, test public key endpoint (it downloads) )
6. Try to configure Tesla custom integration  (i can authenticate but.. i have a persistent error "failed to set up, check the logs")

I repeated these steps over and over again..  

here is a scrap from the Tesla Proxy log (**** for erasing some info):

`[18:25:19]` werkzeug:INFO: 172.30.32.2 - - [14/Nov/2024 18:25:19] "GET / HTTP/1.1" 200 -
[18:27:46] webui:DEBUG: Callback URL: https://***mobile3lounge.duckdns.org/callback?locale=en-US&code=NA_d2e12d7c96e407f7527f8f8756c81******9d95fe38cea828567913dee6&state=dc1afff2f5&issuer=https%3A%2F%2Fauth.tesla.com%2Foauth2%2Fv3
[18:27:46] webui:DEBUG: code: NA_d2e12d7c96e407f7527f8f8******10c4a319d95fe38cea828567913dee6
[18:27:46] urllib3.connectionpool:DEBUG: Starting new HTTPS connection (1): auth.tesla.com:443
[18:27:47] urllib3.connectionpool:DEBUG: https://auth.tesla.com:443 "POST /oauth2/v3/token HTTP/1.1" 200 2270
[18:27:47] webui:WARNING: Obtained refresh token: NA_1f3769c859a84cd4dd3319******4d13ffe4ab04c92590e773e233592ed703ea
[18:27:47] werkzeug:INFO: 172.30.32.2 - - [14/Nov/2024 18:27:47] "GET /callback?callback_url=https://***mobile3lounge.duckdns.org/callback?locale%3Den-US%26code%3DNA_d2e12d7c96e4******7f8f8756c8110c4a319d95fe38cea828567913dee6%26state%3Ddc1afff2f5%26issuer%3Dhttps%253A%252F%252Fauth.tesla.com%252Foauth2%252Fv3 HTTP/1.1" 200 -
2024-11-14T18:28:14+02:00 [info ] Received GET request for /api/1/products
2024-11-14T18:28:14+02:00 [debug] Forwarding request to https://fleet-api.prd.na.vn.cloud.tesla.com/api/1/products
2024-11-14T18:28:17+02:00 [info ] Received GET request for /api/1/products
2024-11-14T18:28:17+02:00 [debug] Forwarding request to https://fleet-api.prd.na.vn.cloud.tesla.com/api/1/products
2024-11-14T18:28:19+02:00 [info ] Received GET request for /api/1/products``

**Reply by hexonMD** [reply]
In case anyone else can't find the public keys that the teslaproxy creates, I was only able to access them through SSH terminal to copy. I had been using studiocodeserver and could never find them before.

**Comment by MicheleMercuri** [comment]
> The old Owner API still works for my 2021 model 3, for reading sensors only. Commands don’t work. I expect they will disable it completely at some like.

for my 2023 model (not highland) commands still works.

**Comment by mheloy** [comment]
Hi,

I've tried to configure using NPM (PRevious NGINX SSL proxy)
I'm getting this Fix Public key before proceding error

[17:05:23] webui:INFO: Starting Flask server for Web UI...
[17:05:23] werkzeug:INFO: WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on all addresses (0.0.0.0)
 * Running on http://127.0.0.1:8099
 * Running on http://172.30.33.7:8099
[17:05:23] werkzeug:INFO: Press CTRL+C to quit
[17:05:28] WARNING: Nginx is not running
[17:05:28] INFO: Generating self-signed SSL certificate
-----
[17:05:28] INFO: Generating keypair
-----BEGIN PUBLIC KEY-----
xxxxxxxxxxxxxxxxxxxxxxxx
-----END PUBLIC KEY-----
[17:05:28] INFO: Testing public key...
[17:07:52] FATAL: Fix public key before proceeding.
s6-rc: info: service legacy-services: stopping

not sure what's going on...
I havve 
HAOS
DuckDNS... but not using letsecrypt
NPM registering both domails melha.duckdns.org meltesla.duckdns.org
   - both proxy host are forwarding to HA IP on port 8123... is this correct?
   - both uses lets encrypt for SSL cert
 
Please advise

**Reply by jerry34ha** [reply]
In my working install, the mytesla.duckdns.org forwards to HA IP on port 4430.

**Reply by mheloy** [reply]
4430 I believe is the port for the Tesla HTTP Proxy... interesting.. I'll keep this mind it can my setup breaks again.

**Comment by mheloy** [comment]
when I type https://mytesla.duckdns.org/.well-known/appspecific/com.tesla.3p.public-key.pem on my 4g phone connected... I can download the PEM file. 

but running of my desktop... I'm not getting any... even 404 error nothing...
maybe my pfsense is doing something different.

any suggestions?


### EDIT ####
Finally its now working after adding a DNS entry to loop mytesla.duckdns.org to the HA IP

[13:18:19] INFO: Found existing keypair
[13:18:19] INFO: Testing public key...
HTTP/2 200 
server: nginx
date: Tue, 14 May 2024 03:18:19 GMT
content-type: application/x-x509-ca-cert
content-length: 178
cache-control: public, max-age=2678400
referrer-policy: no-referrer
x-content-type-options: nosniff
x-frame-options: SAMEORIGIN
etag: "17cf37a775c073f4-b2"
last-modified: Tue, 14 May 2024 01:46:16 GMT
accept-ranges: bytes

-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE4JG57LIOiWoNcwAz8DSC0w736+XH
qRUKum6vpwQ6VSPZsY2QGDU114X/ExDvjUT9TkZvQRhqgJntdwlywv9nrw==
-----END PUBLIC KEY-----
[13:18:19] NOTICE: Request application access with Tesla, then fill in credentials and restart addon.
s6-rc: info: service legacy-services: stopping
s6-rc: info: service legacy-services successfully stopped
s6-rc: info: service webui: stopping
s6-rc: info: service webui successfully stopped
s6-rc: info: service legacy-cont-init: stopping
s6-rc: info: service legacy-cont-init successfully stopped
s6-rc: info: service fix-attrs: stopping
s6-rc: info: service fix-attrs successfully stopped
s6-rc: info: service s6rc-oneshot-runner: stopping
s6-rc: info: service s6rc-oneshot-runner successfully stopped

After this, I had issue with enrolling the car where in its not allowing third party apps... I just had to start from scratch the Tesla Proxy and now working . I can control my charging amps again with my automation on solar output.

**Comment by guyank** [comment]
I had this running using the steps above but was then suffering throttling issues so was trying to get it working again. Now, I can't set it up. When I run through the config flow, it fails to setup, and I get the following message in the log, "ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self-signed certificate (_ssl.c:1000)". I am using Let's Encrypt for my certificate.  Any ideas where to start?

**Reply by llamafilm** [reply]
There are two different certificates here. One for the public key which is internet-facing. The other for internal access only from Home Assistant, which is self signed. If you're seeing this error during the config flow then it sounds like the latter. The addon is supposed to copy that cert to `/share/tesla/selfsigned.pem` where it can be read by the Tesla integration to trust it. I'm guessing that file is still the old one, so you need to delete it and then re-run the addon setup.

**Reply by guyank** [reply]
Thanks. I tried that and was sure that was going to be the cause, because the add-on creates the cert in a different place to where it's used when configuring by NPM. I've tried a couple of times, making sure that the certificate and key are deleted in both places and then starting from scratch (I've even deleted the add-on and started again), but still the same error. Looking at the logs for the add-on, I'm also getting:

2024/08/22 16:30:33 http: TLS handshake error from 172.30.32.1:54590: EOF
[16:33:03] werkzeug:INFO: 172.30.32.2 - - [22/Aug/2024 16:33:03] "GET / HTTP/1.1" 200 -

**Comment by VictorTorrens** [comment]
Hello to everyone, I lost it when i need to move the pem file in homeassistant.
I moved it from share/tesla, to /www/tesla, but i cant reach it.
If someone can give me some insight i will really apreciate it

Thankyou in advance

**Reply by jerry34ha** [reply]
I copied 2 files in /config/www/tesla: selfsigned.pem and com.tesla.3p.public-key.pem

**Comment by jerry34ha** [comment]
Since a few weeks, I now get this message when the Tesla integration starts:
2026-02-01 09:01:58.656 WARNING (MainThread) [custom_components.tesla_custom] Unable to load custom SSL certificate from  /config/www/tesla/selfsigned.pem

I struggled to find out why, but the reason was simple: I had to re-authenticate:

* Open Tesla HTTP proxy add interface
* Login to Tesla
* Copy the call back URL
* Generate the Token

Then go back to the Tesla integration:
* reconfigure
* Select Fleet API
* Copy token to the "refresh token" field
* update the path to the certificate to /config/www/tesla/selfsigned.pem
* Save.
